crates/ruff/src/rules/ruff/rules/invalid_index_type.rs

```diff
@@ -0,0 +1,186 @@
+                            checker.diagnostics.push(Diagnostic::new(
+                                InvalidIndexType {
+                                    var_type: expression_type_name(value)
+                                        .expect("Failed to cast parent expression to type name")
+                                        .to_string(),
+                                    idx_type: constant_type_name(index_value).to_string(),
+                                    slice_bound: true,
+                                },
+                                slice_bound.range(),
+                            ));
+                        }
+                    } else if matches!(
+                        slice_bound.as_ref(),
+                        Expr::Tuple(_)
+                            | Expr::List(_)
+                            | Expr::Set(_)
+                            | Expr::Dict(_)
+                            | Expr::ListComp(_)
+                            | Expr::SetComp(_)
+                            | Expr::DictComp(_)
+                            | Expr::JoinedStr(_)
+                    ) {
+                        checker.diagnostics.push(Diagnostic::new(
+                            InvalidIndexType {
+                                var_type: expression_type_name(value)
+                                    .expect("Failed to cast parent expression to type name")
+                                    .to_string(),
+                                idx_type: expression_type_name(slice_bound)
+                                    .expect("Failed to slice bound expression to type name")
+                                    .to_string(),
+                                slice_bound: true,
+                            },
+                            slice_bound.range(),
+                        ));
+                    }
+                }
+            }
+            // If it's some other data type, it's a violation
+            Expr::Tuple(_)
+            | Expr::List(_)
+            | Expr::Set(_)
+            | Expr::Dict(_)
+            | Expr::ListComp(_)
+            | Expr::SetComp(_)
+            | Expr::DictComp(_)
+            | Expr::JoinedStr(_) => {
+                checker.diagnostics.push(Diagnostic::new(
+                    InvalidIndexType {
+                        var_type: expression_type_name(value)
+                            .expect("Failed to cast parent expression to type name")
+                            .to_string(),
+                        idx_type: expression_type_name(slice)
+                            .expect("Failed to cast index expression to type name")
+                            .to_string(),
+                        slice_bound: false,
+                    },
+                    slice.range(),
+                ));
+            }
+            // If it's anything else, it's too hard to tell if it's a violation
+            _ => (),
```

Comment:
Should we list the expressions here to force a compile error when a new expression was added.

Reply:
I don't think so. Unless new literal types are introduced, a new expression is unlikely to be detectable by this rule as a violation.